### DISCUSSION TITLE
AO map not included when auto converting to USDZ

### DISCUSSION THREAD

**Discussion by danilokrulj** [google/model-viewer#4816] [Q&A]
So I have a model that needs AO to look decent but I want to tile the other textures (base,normal,rough).

I add the AO map to the occlusionTexture channel through code and its visible in the 3d viewer and in AR on Android. The problem is that on iOS the map isn't shown in AR, I'm guessing it gets lost during the USDZ auto-conversion.

I tried exporting the model with AO from model-viewer and than tried using that as the original model but again same result.
Also tried importing the exported glb into Blender to see what node it adds for the occlusion texture, exported that model and tried using that as the original and again same result.

I guess I could convert it manually but that doesn't work for me because the model has lots of variations which I change through code and its simply absurd to make all corresponding usdz versions.

Am I missing something or does USDZ conversion just not support AO from gltf? Any help is welcome! :)

EDIT:

I manually converted the model to USDZ and saw that the occlusion texture is present in the usdz, than i placed that usdz in my ios-src and again there was no AO in AR, does that mean this is a QuickLook issue?

**Comment by elalish** [comment]
It certainly sounds like a Quick Look issue - could you attach your GLB and USDZ models and perhaps some screen shots? Hopefully @AdaRoseCannon can take a look.

**Comment by AdaRoseCannon** [comment]
I spent the morning testing this.

I created a model in Blender with a test Ambient Occlusion Map I saved it out as USDz using the THREE.js editor. 

The USDz file appears to have the AO map as expected. It's visible in both finder and preview on my mac.

I also tried sending it to my Vision Pro and my iPhone and it worked as expected on both of these. 

Please can you let me know your iPhone version, and I will look into it, I'm definitely not the expert on Quick Look so filing a bug with Feedback Assistant would probably get the most visibility with the right people.

**Reply by danilokrulj** [reply]
The devices I tested on are: 

Iphone XR iOS 17.4.1
Iphone 12 iOS 17.4.1
Iphone 8 iOS 16.7.8

Using the latest model viewer version.

I cant send the models though, since the project is still in the works. :/
But the model just has a BaseColor, Normal and Metallic&Roughness map. I add the AO map through code to the occlusionTexture channel. One thing to note is that all the maps have tiling accept the AO map.
This is how I apply the AO texture, maybe i'm doing something wrong there?

`var scale1 = {u:3, v:3};

const createAndApplyTexture = async (materialToChange, material) => {
  const texture_baseColor = await modelViewer.createTexture(material.base_url);
  const texture_normal = await modelViewer.createTexture(material.normal_url);
  const texture_roughness = await modelViewer.createTexture(material.roughness_url);

  materialToChange.pbrMetallicRoughness.baseColorTexture.setTexture(texture_baseColor);
  materialToChange.normalTexture.setTexture(texture_normal);
  materialToChange.pbrMetallicRoughness.metallicRoughnessTexture.setTexture(texture_roughness);


  sampler = materialToChange.pbrMetallicRoughness.baseColorTexture.texture.sampler;
  samplerNormal = materialToChange.normalTexture.texture.sampler;
  sampler.setScale(scale1);
  samplerNormal.setScale(scale1);
}

async function createAOTexture(){
  texture1_1 = await modelViewer.createTexture("mat1_AO.png");
  mat1.occlusionTexture.setTexture(texture1_1);
}`

**Reply by elalish** [reply]
There's not much we can do to help without a repro. I'd recommend creating a simpler model that has the same problem and isn't secret.

**Reply by danilokrulj** [reply]
I think the issue here was that I wanted to use the AO map with UV wrap and tile the rest of the textures (I tried doing this again today and ran into this old discussion), which appears to work in the viewer and on android but doesn't work in QuickLook USDZ. I tried this again today and figured out that when the model gets converted the AO texture appears to be merged and tiled with the base color texture and my model looks something like the model in this discussion: 
https://github.com/google/model-viewer/discussions/4392#discussion-5450606

So I guess the answer is that USDZ doesn’t support multiple UV channels? I know I can bake the AO in to the base color texture but the model is kind of large and I was trying to save up space and loading time while keeping the details :)

**Reply by elalish** [reply]
Agreed, that sounds like a deficiency of USDZ - you're welcome to generate the USDZ yourself and use `ios-src` so you don't have to bloat your GLB. @AdaRoseCannon in case anyone at Apple wants to address this in USDZ.
